India NHM Data: State-wise and District-wise data

SenthamizhanV opened the pull request:
@pradh @thejeshgn 

Added National Health Mission data for state-wise maternal data under `india_nhm` folder.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/datacommonsorg/data/414)
<!-- Reviewable:end -->

pradh reviewed (changes requested):
Good start!

> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/data/414#-:-MYG2FxB3EhF-mdUe9G1:beulzst)* status: 0 of 24 files reviewed, 7 unresolved discussions (waiting on @SenthamizhanV)

___
*[scripts/india_nhm/base/data_cleaner.py, line 89 at r1](https://reviewable.io/reviews/datacommonsorg/data/414#-MYFys2dATjGrEy-hO-i:-MYFys2dATjGrEy-hO-j:b-gia0n4) ([raw file](https://github.com/datacommonsorg/data/blob/f2cbf37595197358a4b2f2b6a3317dd1ba72de5c/scripts/india_nhm/base/data_cleaner.py#L89)):*
> ```
> ···
> ```

please consider running pylint/yapf: https://github.com/datacommonsorg/data#guidelines
___
*[scripts/india_nhm/base/data_cleaner.py, line 105 at r1](https://reviewable.io/reviews/datacommonsorg/data/414#-MYFz2hK7ncIvI69De7N:-MYFz2hK7ncIvI69De7O:brts24e) ([raw file](https://github.com/datacommonsorg/data/blob/f2cbf37595197358a4b2f2b6a3317dd1ba72de5c/scripts/india_nhm/base/data_cleaner.py#L105)):*
> ```
> 3:
> ```

Please add a comment on what's from column 3 onwards, and is that true for all NHM CSVs?
___
*[scripts/india_nhm/base/data_cleaner.py, line 123 at r1](https://reviewable.io/reviews/datacommonsorg/data/414#-MYFzXBP3WAr6hYb-_5v:-MYFzXBP3WAr6hYb-_5w:b-tq7kyi) ([raw file](https://github.com/datacommonsorg/data/blob/f2cbf37595197358a4b2f2b6a3317dd1ba72de5c/scripts/india_nhm/base/data_cleaner.py#L123)):*
> ```
> '20', fname[-2:]
> ```

Any reason this shouldn't be the first 4 letters of fname (i.e, why hard code '20') ?
___
*[scripts/india_nhm/base/data_cleaner.py, line 125 at r1](https://reviewable.io/reviews/datacommonsorg/data/414#-MYFzd9Q2ItNI6pPYMC-:-MYFzd9Q2ItNI6pPYMC0:b-2k75g6) ([raw file](https://github.com/datacommonsorg/data/blob/f2cbf37595197358a4b2f2b6a3317dd1ba72de5c/scripts/india_nhm/base/data_cleaner.py#L125)):*
> ```
> if fext == '.xls':
> ```

else: error?
___
*[scripts/india_nhm/base/data_cleaner.py, line 137 at r1](https://reviewable.io/reviews/datacommonsorg/data/414#-MYG-J7tD4gK6kGmwzku:-MYG-J7tD4gK6kGmwzkv:b4u92f4) ([raw file](https://github.com/datacommonsorg/data/blob/f2cbf37595197358a4b2f2b6a3317dd1ba72de5c/scripts/india_nhm/base/data_cleaner.py#L137)):*
> ```
> 2:
> ```

Comment on what is 2 onwards?
___
*[scripts/india_nhm/maternal_health/NHM_MaternalHealth.tmcf, line 7 at r1](https://reviewable.io/reviews/datacommonsorg/data/414#-MYG1ayD3J8GydtBGMmO:-MYG1ayD3J8GydtBGMmP:b-t99e71) ([raw file](https://github.com/datacommonsorg/data/blob/f2cbf37595197358a4b2f2b6a3317dd1ba72de5c/scripts/india_nhm/maternal_health/NHM_MaternalHealth.tmcf#L7)):*
> ```
> dcs:Count_PregnancyEvent
> ```

These variables lack definition at the moment. That is fine, but could do the following:

1. Prefix the DCID with "indianNHM/", so it is indianNHM/Count_PregnancyEvent.

2. Define StatisticalVariable nodes of the form:

```
Node: dcid:indianNHM/Count_PregnancyEvent
typeOf: dcs:StatisticalVariable
populationType: schema:Person
measuredProperty: dcs:indianCensus/Count_PregnancyEvent
```

For example, see: https://github.com/datacommonsorg/data/blob/master/scripts/india_census/primary_religion_data/IndiaCensus2011_Primary_Abstract_Religion.mcf
___
*[scripts/india_nhm/maternal_health/README.md, line 27 at r1](https://reviewable.io/reviews/datacommonsorg/data/414#-MYG1DSiB-EGBbDpJ_Ge:-MYG1DSiB-EGBbDpJ_Gf:b18kmre) ([raw file](https://github.com/datacommonsorg/data/blob/f2cbf37595197358a4b2f2b6a3317dd1ba72de5c/scripts/india_nhm/maternal_health/README.md#L27)):*
<details><summary><i alt="summary">Quoted 12 lines of code…</i></summary>

> ```
>  State: State
>  isoCode: isoCode
>  Date: Date
>  Count_PregnancyEvent: Estimated Number of Annual Pregnancies #
>  Count_PregnantWomen_RegisteredForAntenatalCare: Total number of pregnant women Registered for ANC
>  Count_PregnantWomen_RegisteredForAntenatalCareWithinFirstTrimester: Number of Pregnant women registered within first trimester
>  Count_ChildDeliveryEvent: Total reported deliveries
>  Count_ChildDeliveryEvent_InAnInstitution: Institutional deliveries (Public Insts.+Pvt. Insts.)
>  Count_ChildDeliveryEvent_InPublicInstitution: Deliveries Conducted at Public Institutions
>  Count_ChildDeliveryEvent_AtHome: Number of Home deliveries
>  Count_ChildDeliveryEvent_AtHome_WithStandByAssist: Number of home deliveries attended by SBA trained (Doctor/Nurse/ANM)
>  Count_DeliveryEvent_Safe_AsFractionOf_Count_DeliveryEvent: % Safe deliveries to Total Reported Deliveries
> ```
</details>


If you view this file this part is not very readable. Maybe you want a "- " before it or something.


<!-- Sent from Reviewable.io -->

pradh reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/data/414)* status: 0 of 24 files reviewed, 9 unresolved discussions (waiting on @SenthamizhanV)

___
*[scripts/__init__.py, line 1 at r1](https://reviewable.io/reviews/datacommonsorg/data/414#-MYG3Wrf6rrzpcOh_9HL:-MYG3Wrf6rrzpcOh_9HM:b-lpruas) ([raw file](https://github.com/datacommonsorg/data/blob/f2cbf37595197358a4b2f2b6a3317dd1ba72de5c/scripts/__init__.py#L1)):*
> ```
> # -*- coding: utf-8 -*-
> ```

For the py files, please add a license like: https://github.com/datacommonsorg/data/blob/master/scripts/india_census/primary_religion_data/preprocess.py#L1-L13
___
*[scripts/__init__.py, line 5 at r1](https://reviewable.io/reviews/datacommonsorg/data/414#-MYG3i9NAdnYVe0Zfy2n:-MYG3i9NAdnYVe0Zfy2o:b-q7zyy0) ([raw file](https://github.com/datacommonsorg/data/blob/f2cbf37595197358a4b2f2b6a3317dd1ba72de5c/scripts/__init__.py#L5)):*
> ```
> @author: Ryzen
> ```

Add your name?


<!-- Sent from Reviewable.io -->

SenthamizhanV commented:
> Any reason this shouldn't be the first 4 letters of fname (i.e, why hard code '20') ?

The first four digits are the previous year's value. Hence I'm only taking the last two digits and adding '20' in front of them. I can take the first four digits and add one to them but I thought this looks clean

> > if fext == '.xls':
>
> else: error?

It parses only if it is an .xls error, else it skips the file.

 -----------------------
**I have implemented the suggested changes and updated the PR. Kindly take a look at it.**

pradh reviewed (approved):
Looking good. Just a few last comments.

> Reviewed 1 of 24 files at r1, 2 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/datacommonsorg/data/414#-:-MYi5tEv9uoUii820dsC:bd10qze)* status: 3 of 26 files reviewed, 15 unresolved discussions (waiting on @SenthamizhanV)

___
*[scripts/india_nhm/base/data_cleaner.py, line 125 at r1](https://reviewable.io/reviews/datacommonsorg/data/414#-MYFzd9Q2ItNI6pPYMC-:-MYi34k55dTosboRwSRS:btqurrt) ([raw file](https://github.com/datacommonsorg/data/blob/f2cbf37595197358a4b2f2b6a3317dd1ba72de5c/scripts/india_nhm/base/data_cleaner.py#L125)):*
<details><summary><i>Previously, pradh (Prashanth R) wrote…</i></summary><blockquote>

> ```
> if fext == '.xls':
> ```

else: error?
</blockquote></details>

Is this comment to handle error in the else case handled?
___
*[scripts/india_nhm/base/data_cleaner.py, line 99 at r2](https://reviewable.io/reviews/datacommonsorg/data/414#-MYi3jLuERurJihpmeKn:-MYi3jLv6H_wSu__1TDl:b-62s6hi) ([raw file](https://github.com/datacommonsorg/data/blob/1ed9fe3dd602c80756aad41c458f83c09f5f6333/scripts/india_nhm/base/data_cleaner.py#L99)):*
> ```
> easuredProperty: dcs:indianNHM/{statvar}
> ```
Sorry, missed this, could you also add:

statType: dcs:measuredValue
___
*[scripts/india_nhm/maternal_health/NHM_MaternalHealth.mcf, line 2 at r2](https://reviewable.io/reviews/datacommonsorg/data/414#-MYi3Y7u1fFiJAZBSdNA:-MYi3Y7u1fFiJAZBSdNB:b-jqrxhl) ([raw file](https://github.com/datacommonsorg/data/blob/1ed9fe3dd602c80756aad41c458f83c09f5f6333/scripts/india_nhm/maternal_health/NHM_MaternalHealth.mcf#L2)):*
> ```
>  #
> ```
Remove the # ?
___
*[scripts/india_nhm/maternal_health/NHM_MaternalHealth.mcf, line 8 at r2](https://reviewable.io/reviews/datacommonsorg/data/414#-MYi4-rt0NvopFMg2n78:-MYi4-rt0NvopFMg2n79:btolekl) ([raw file](https://github.com/datacommonsorg/data/blob/1ed9fe3dd602c80756aad41c458f83c09f5f6333/scripts/india_nhm/maternal_health/NHM_MaternalHealth.mcf#L8)):*
> ```
> Total number of pregnant women Registered for ANC
> ```

Expand ANC.  Also, please be consistent with the use capitals.  The description can just be: "Total number of pregnant women registered for antenatal care."
___
*[scripts/india_nhm/maternal_health/NHM_MaternalHealth.mcf, line 50 at r2](https://reviewable.io/reviews/datacommonsorg/data/414#-MYi4O6E3w8cZ2Y8NZLV:-MYi4O6F4nb8DH1P7fj7:bekh2zs) ([raw file](https://github.com/datacommonsorg/data/blob/1ed9fe3dd602c80756aad41c458f83c09f5f6333/scripts/india_nhm/maternal_health/NHM_MaternalHealth.mcf#L50)):*
> ```
> %
> ```

Percent of 
___
*[scripts/india_nhm/maternal_health/NHM_MaternalHealth.tmcf, line 7 at r1](https://reviewable.io/reviews/datacommonsorg/data/414#-MYi5A1Y4F05uU_cF_sT:-MYi5A1Y4F05uU_cF_sU:bknc5pr) ([raw file](https://github.com/datacommonsorg/data/blob/f2cbf37595197358a4b2f2b6a3317dd1ba72de5c/scripts/india_nhm/maternal_health/NHM_MaternalHealth.tmcf#L7)):*
> ```
> Count_PregnancyEvent
> ```
This should now match the DCID we defined for the stat-var, indianNHM/Count_PregnancyEvent.

The idea is that when we fully define the StatisticalVariable node, we will remove the prefix.
___
*[scripts/india_nhm/maternal_health/test/test_gen.csv, line 1 at r2](https://reviewable.io/reviews/datacommonsorg/data/414#-MYi5hHJ2dnYHFsLz_1_:-MYi5hHJ2dnYHFsLz_1a:b-bjcxyq) ([raw file](https://github.com/datacommonsorg/data/blob/1ed9fe3dd602c80756aad41c458f83c09f5f6333/scripts/india_nhm/maternal_health/test/test_gen.csv#L1)):*
> ```CSV
> State,isoCode,Date,Count_PregnancyEvent,Count_PregnantWomen_RegisteredForAntenatalCare,Count_PregnantWomen_RegisteredForAntenatalCareWithinFirstTrimester,Count_ChildDeliveryEvent,Count_ChildDeliveryEvent_InAnInstitution,Count_ChildDeliveryEvent_InPublicInstitution,Count_ChildDeliveryEvent_AtHome,Count_ChildDeliveryEvent_AtHome_WithStandByAssist,Count_DeliveryEvent_Safe_AsFractionOf_Count_DeliveryEvent
> ```

Should this file be checked in?


<!-- Sent from Reviewable.io -->

SenthamizhanV commented:
@pradh I'm trying to upload the other three datasets but the checks are failing. I ran pylint and yapf multiple times but its still failing. And the Cloud Build shows I do not have necessary permissions. Can you help me finding where the issue is?

pradh commented:
> @pradh I'm trying to upload the other three datasets but the checks are failing. I ran pylint and yapf multiple times but its still failing. And the Cloud Build shows I do not have necessary permissions. Can you help me finding where the issue is?

Tests are failing (as below). Added you as a logs viewer, can you see this: https://pantheon.corp.google.com/cloud-build/builds/c5af7588-1413-4832-b100-4548d958a354;step=1?project=datcom-ci ?

```
======================================================================
ERROR: india_nhm.nhm_maternal_health.preprocess_test (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: india_nhm.nhm_maternal_health.preprocess_test
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/unittest/loader.py", line 436, in _find_test_path
    module = self._get_module_from_name(name)
  File "/usr/local/lib/python3.7/unittest/loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "/workspace/scripts/india_nhm/nhm_maternal_health/preprocess_test.py", line 21, in <module>
    from base.data_cleaner import NHMDataLoaderBase
ModuleNotFoundError: No module named 'base'
```

SenthamizhanV commented:
@pradh I had that issue when I ran pylint locally. And I rectified it. I'm not sure why it is not working when pushed. I will look into it. And I still couldn't see the logs in Cloud Build. It shows my commits now but it still says `Error: The caller does not have permission`. And this https://pantheon.corp.google.com/cloud-build/builds/c5af7588-1413-4832-b100-4548d958a354;step=1?project=datcom-ci link asks for sign-in credentials with @google.com email.

pradh commented on scripts/__init__.py:
```diff
@@ -0,0 +1,6 @@
+Created on Sat Apr 10 20:04:47 2021
+
+@author: Ryzen
+"""
```
Remove this file?

pradh commented on scripts/india_nhm/README.md:
```diff
@@ -0,0 +1,5 @@
+# NHM Datasets
```
Can you not check-in the .pylint.d directory?

pradh commented on scripts/india_nhm/nhm_maternal_health/NHM_MaternalHealth.mcf:
```diff
@@ -1,54 +1,63 @@
 
 Node: dcid:indianNHM/Count_PregnantWomen_RegisteredForAntenatalCare
-description: "Total number of pregnant women Registered for ANC"
+description: "Total number of pregnant women registered for anc"
```
Can you expand "anc"? Its an abbreviation right

pradh commented on scripts/india_nhm/nhm_maternal_health/NHM_MaternalHealth.mcf:
```diff
@@ -1,54 +1,63 @@
 Node: dcid:indianNHM/Count_PregnancyEvent
-description: "Estimated Number of Annual Pregnancies #"
+description: "Estimated number of annual pregnancies #"
```
Drop # (same comment as before)?

pradh commented on scripts/india_nhm/base/data_cleaner.py:
```diff
@@ -84,19 +84,20 @@
 observationDate: C:{dataset_name}->Date
 observationPeriod: "P1Y"
-value: C:{dataset_name}->{statvar}
+value: C:{dataset_name}->indianNHM/{statvar}
```
For the column name, adding the indianNHM/ seems incorrect ?

SenthamizhanV commented:
@pradh Can you merge this PR for india_nhm? This is the version with both districts and states.
